lib/matplotlib/cbook/__init__.py

```diff
@@ -1537,6 +1565,115 @@ def pts_to_prestep(x, *args):
+        length ``N``, each of these arrays will be length ``2N - 2``. For
+        ``N=0``, the length will be 0.
+
+    """
+    args = np.array(args)
+    step_length = max(2 * max(len(x), len(args[0])) - 2, 0)
+    steps = np.zeros((1 + len(args), step_length))
+
+    if len(x) == len(args[0]) + 1:
+        steps[0, ::len(steps[0])-1] = x[::len(x)-1]
+        steps[0, 2::2] = x[1:-1]
+        steps[0, 1:-1:2] = x[1:-1]
+        steps[1:, 0::2] = args
+        steps[1:, 1::2] = args
+
+    elif len(x) + 1 == len(args[0]):
+        steps[0, 0::2] = x
+        steps[0, 1::2] = x
+        steps[1:, ::len(steps[0])-1] = args[:, ::len(args[0])-1]
+        steps[1:,  1:-1:2] = args[:, 1:-1]
+        steps[1:,  2:-1:2] = args[:, 1:-1]
+    else:
+        # Fall back to steps-post (legend drawing)
+        steps = pts_to_poststep(x, *args)
+
+    return steps
+
+
+def pts_to_betweenstep_edges(x, *args):
+    """
+    Convert continuous line to between-steps, adding edges
+
+    Given a set of ``N`` edges and ``N - 1`` values padded with Nan to size N,
+    converts to ``2N`` points, which when connected linearly give
+    a step function connecting step edges at a specified value, with the first
+    and last edge going to 0
+
+    Parameters
+    ----------
+    x : array
+        The x location of the step edges. May be empty.
+
+    y1, ..., yp : array
+        y arrays to be turned into steps; must have length as ``x`` +/- 1.
+    Returns
+    -------
+    out : array
+        The x and y values converted to steps in the same order as the input;
+        can be unpacked as ``x_out, y1_out, ..., yp_out``.  If the input is
+        length ``N``, each of these arrays will be length ``2N``. For
+        ``N=0``, the length will be 0.
+
+    """
+    steps = pts_to_betweenstep(x, *args)
+
+    # Extra steps to plot edges where values are missing (Nan).
+    nan_cols = np.nonzero(np.isnan(np.sum(steps, axis=0)))[0]
+    nan_cols[1::2] = nan_cols[1::2]+1
+    pad_steps = []
+
+    xlike = len(x) == len(args[0]) + 1
+    ylike = len(x) + 1 == len(args[0])
+    if not (xlike or ylike):
+        # Fall back to steps-post (legend drawing)
+        edge_steps = pts_to_poststep(x, *args)
+        return edge_steps
+
+    for part in np.split(steps, nan_cols, axis=1):
+        if not np.isnan(np.sum(part)):
+            pad_part = np.zeros((2, part.shape[-1]+2))
+            pad_part[:, 1:-1] = part
+            if xlike:
+                pad_part[:, 0] = part[:, 0]
+                pad_part[1, 0] = 0
+                pad_part[:, -1] = part[:, -1]
+                pad_part[1, -1] = 0
+            else:
+                pad_part[:, 0] = 0
+                pad_part[1, 0] = part[1, 0]
+                pad_part[:, -1] = 0
+                pad_part[1, -1] = part[1, -1]
+
+            pad_steps.append(pad_part)
+        else:
+            pad_steps.append(part)
+
+    return np.hstack(pad_steps)
+
```

Comment:
It really seems like all these functions could be generalized a it better, or use each other rather than using similar code over and over.

Reply:
Maybe it could, however, a lot of the complexity was necessary for step (especially edges) to be well behaved with Nans. Otherwise `edges` is using the `between` code.
```
ax.step([1,2,3,4], [1,np.nan,3], where='edges')
ax.step([1,2,np.nan,4], [1,2,3,4,5], where='edges')
```
![image](https://user-images.githubusercontent.com/13226500/71690844-22ce7c00-2da6-11ea-8989-56c25de057fd.png)
![image](https://user-images.githubusercontent.com/13226500/71690877-32e65b80-2da6-11ea-8d9b-4320beff678d.png)